frontend/src/routes/ProtectedRoute.tsx

```diff
@@ -0,0 +1,21 @@
+import { Navigate, useLocation } from "react-router-dom";
+import { useSupabaseSession } from "../hooks/useSupabaseSession";
+
+interface ProtectedRouteProps {
+  element: React.ReactNode;
+}
+
+const ProtectedRoute: React.FC<ProtectedRouteProps> = ({ element }) => {
+  const location = useLocation();
+  const { session, loading } = useSupabaseSession();
+
+  if (loading) return null;
+
+  return session ? (
+    <>{element}</>
+  ) : (
+    <Navigate to="/login" state={{ from: location.pathname }} replace />
```

Comment:
I want to ask why you used <Navigate>.  This document says to use `useNavigate` is common. 

> Having a component-based version of the useNavigate hook makes it easier to use this feature in a [React.Component](https://reactjs.org/docs/react-component.html) subclass where hooks are not able to be used.

https://reactrouter.com/en/main/components/navigate

Reply:
It seems ProtectedRoute is used as a JSX component in App.tsx.b This means the ProtectedRoute should return Element so if it is used useNavigate here, it will return void instead of Element.

```js
          <Route
            path="/expenses/friendslist"
            element={<ProtectedRoute element={<FriendsListPage />} />}
          />
```

Reply:
In the future, we may use this react-router-dom feature。

https://reactrouter.com/en/main/route/error-element